tests/raftstore/pd.rs

```diff
@@ -82,114 +99,357 @@ impl Cluster {
+            // No range covers [start, end) now, insert directly.
+            self.add_region(&region);
+        } else {
+            // overlap, remove old, insert new.
+            // E.g, 1 [a, c) -> 1 [a, b) + 2 [b, c), either new 1 or 2 reports, the region
+            // is overlapped with origin [a, c).
+            assert!(version > search_version);
+            assert!(conf_ver >= search_conf_ver);
+
+            self.remove_region(&search_region);
+            self.add_region(&region);
+        }
 
         Ok(())
     }
 
-    fn split_region(&mut self, left: metapb::Region, right: metapb::Region) -> Result<()> {
-        let left_end_key = enc_end_key(&left);
-        let right_end_key = enc_end_key(&right);
-
-        assert!(left.get_start_key() < left.get_end_key());
-        assert_eq!(left.get_end_key(), right.get_start_key());
-        assert!(right.get_end_key().is_empty() || right.get_start_key() < right.get_end_key());
+    fn handle_heartbeat_conf_ver(&mut self,
+                                 region: metapb::Region,
+                                 leader: metapb::Peer)
+                                 -> Result<pdpb::RegionHeartbeatResponse> {
+        let conf_ver = region.get_region_epoch().get_conf_ver();
+        let end_key = enc_end_key(&region);
 
-        // origin pre-split region's end key is the same as right end key,
-        // and must exists.
-        if !self.regions.contains_key(&right_end_key) {
-            return Err(box_err!("region {:?} doesn't exist", right));
+        let cur_region = self.get_region_by_id(region.get_id()).unwrap();
+
+        let cur_conf_ver = cur_region.get_region_epoch().get_conf_ver();
+        try!(check_stale_region(&cur_region, &region));
+
+        let region_peer_len = region.get_peers().len();
+        let cur_region_peer_len = cur_region.get_peers().len();
+
+        if conf_ver > cur_conf_ver {
+            // If ConfVer changed, TiKV has added/removed one peer already.
+            // So pd and TiKV can't have same peer count and can only have
+            // only one different peer.
+            // E.g, we can't meet following cases:
+            // 1) pd is (1, 2, 3), TiKV is (1)
+            // 2) pd is (1), TiKV is (1, 2, 3)
+            // 3) pd is (1, 2), TiKV is (3)
+            // 4) pd id (1), TiKV is (2, 3)
+
+            assert!(region_peer_len != cur_region_peer_len);
+
+            if cur_region_peer_len > region_peer_len {
+                // must pd is (1, 2), TiKV is (1)
+                assert_eq!(cur_region_peer_len - region_peer_len, 1);
+                let peers = setdiff_peers(&cur_region, &region);
+                assert_eq!(peers.len(), 1);
+                assert!(setdiff_peers(&region, &cur_region).is_empty());
+            } else {
+                // must pd is (1), TiKV is (1, 2)
+                assert_eq!(region_peer_len - cur_region_peer_len, 1);
+                let peers = setdiff_peers(&region, &cur_region);
```

Comment:
yes, and we usually use different as the function name for set complement. 
Suggested by @disksing